grin/src/miner.rs

```diff
@@ -137,14 +139,107 @@ impl Miner {
 			tx_pool: tx_pool,
 			debug_output_id: String::from("none"),
 		}
-	}
+	} // fn new
 
 	/// Keeping this optional so setting in a separate function
 	/// instead of in the new function
 	pub fn set_debug_output_id(&mut self, debug_output_id: String) {
 		self.debug_output_id = debug_output_id;
 	}
 
+	// Get a solution (BlockHeader) from the stratum client
+	fn stratum_get_solution(
+		&self,
+		stream: &mut TcpStream,
+		stream_err: &mut bool,
+	) -> Option<BlockHeader> {
+		// Get a solved block header from the stream
+		// get size
+		let dsz: usize;
+		let mut data = [0 as u8; 4]; // using 4 byte buffer
+		loop {
+			match stream.read_exact(&mut data) {
+				Ok(_) => unsafe {
+					let sz = mem::transmute::<[u8; 4], u32>(data);
+					dsz = sz as usize;
+					break;
+				},
+				Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
+					// This is not an 'error', just isnt any data ready for us to read right now
+					return None;
+				}
+				Err(e) => {
+					// We lost our client
+					warn!(LOGGER, "Error in connection with stratum client: {}", e);
+					*stream_err = true;
+					return None;
+				}
+			}
+		} // loop
+
+		// get serialized data
+		let mut retry = 0;
+		let mut dvec = vec![0; dsz];
+		loop {
+			match stream.read_exact(&mut dvec) {
+				Ok(_) => {
+					let mut bh: BlockHeader = ser::deserialize(&mut &dvec[..]).unwrap();
+					return Some(bh);
+				}
+				Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
+					// Sleep a bit and try again, but give up at some point
+					retry += 1;
+					if retry >= 1000 {
+						warn!(LOGGER, "Error in connection with stratum client: {}", e);
+						*stream_err = true;
+						return None;
+					}
+					thread::sleep(Duration::from_millis(100));
+				}
+				Err(e) => {
+					warn!(LOGGER, "Error in connection with stratum client: {}", e);
+					*stream_err = true;
+					return None;
+				}
+			} // ? {}
+		}
+	} // End get_solution
+
+	// Send a new block to the stratum client
+	fn stratum_send_block(&self, stream: &mut TcpStream, b: &Block, stream_err: &mut bool) {
+		// Serialize the block for sending over the wire
+		let mut bvec = Vec::new();
+		ser::serialize(&mut bvec, &b).expect("serialization failed");
+
+		// Send the size of the block
+		let bveclen: u32 = bvec.len() as u32;
+		unsafe {
+			let bszb = mem::transmute::<u32, [u8; 4]>(bveclen);
+			match stream.write(&bszb) {
+				Ok(_) => {
+					info!(LOGGER, "Sent Block Size");
+				}
+				Err(e) => {
+					warn!(LOGGER, "Error in connection with stratum client: {}", e);
+					*stream_err = true;
+					return;
+				}
+			}
+		}
+
+		// Send block to client
+		match stream.write(&bvec) {
+			Ok(_) => {
+				info!(LOGGER, "Sent block");
+			}
+			Err(e) => {
+				warn!(LOGGER, "Error in connection with stratum client: {}", e);
+				*stream_err = true;
+				return;
+			}
+		}
+	} // End send_block
+
```

Comment:
If possible, I think it would be best to put as much of the stratum-server related code as possible into a separate file to keep them distinct. miner.rs is very messy already (for reasons I'll outline in another issue) and needs cleaning up.